### PR TITLE
Support ~ in filename parsing

### PR DESCRIPTION
**Pull request by bwidawsk**
Untested on non-Linux platforms...

Signed-off-by: Ben Widawsky <ben@bwidawsk.net>

### CODE REVIEW

File: kernel/yosys.cc [hunk 1]
```diff
@@ -651,6 +651,8 @@ void rewrite_filename(std::string &filename)
 		filename = filename.substr(1, GetSize(filename)-2);
 	if (filename.substr(0, 2) == "+/")
 		filename = proc_share_dirname() + filename.substr(2);
+	if (filename.substr(0, 1) == "~")
+		filename = filename.replace(0, 1, getenv("HOME"));
```
Comment (cliffordwolf) [inline kernel/yosys.cc]: The convention is that `~/` expands to `$HOME` and `~<user>/` expands to the home directory of the given user. Either you support only the first case, then the match should be for `~/`, not just `~`, or you support the latter case as well but then you'd have to actoally parse whatever is between the `~` and the first `/` and look up that users home directory (using `getpwnam()` or `getpwnam_r()`).

Comment (eddiehung) [reply kernel/yosys.cc]: Also it might be worth wrapping it around `#ifdef _WIN32`

Comment (bwidawsk) [reply kernel/yosys.cc]: @cliffordwolf Let's start with the first case... and then maybe someone else who cares about the second case can do it.

@eddiehung Good idea (but I think you mean ifndef).

I'll fix it up a bit later.

**Comment by udif** [comment]
I spent 15 minutes of my life to see how ~username is done in bash.

It boils down to:

#include <sys/types.h>
#include <pwd.h>
struct passwd *user_entry;
user_entry = getpwnam (username);
// use user_entry->pw_dir to get user's home dir.

**Pull request merged**